[OpenEmbedded] Fix include and lib paths for multilib targets

mikecrowe opened the pull request:
The fix in e9ddc44a604318739a27249d267aa7267d13d93b wasn't enough for multilib targets since they use a custom vendor that contains the multilib variant.  There are two common multilib variants: lib32 and libx32, so let's support both.

mikecrowe committed:
[OpenEmbedded] Fix include and lib paths for multilib targets

The fix in e9ddc44a604318739a27249d267aa7267d13d93b wasn't enough for
multilib targets since they use a custom vendor that contains the
multilib variant.  There are two common multilib variants: lib32 and
libx32, so let's support both.

mikecrowe commented:
This works, but I'm not sure that it's the best solution. I could teach `parseVendor` to match a prefix of `oeml` rather than hard-coding the two specific multilib types.

mikecrowe commented:
@nikic @MaskRay Ping

MaskRay commented:
The include and library path code in the driver needs more love. I have cleaned up some a few years ago but there is still quite a bit tech debt.

In particular, the 2018 commit you mentioned (https://reviews.llvm.org/D48862) introduced
```
  if (Triple.getVendor() == llvm::Triple::OpenEmbedded &&
      Triple.isArch64Bit())
    addPathIfExists(D, SysRoot + "/usr/" + OSLibDir, Paths);
  else
    addPathIfExists(D, SysRoot + "/usr/lib/../" + OSLibDir, Paths);
```

which could be simplified to `/usr/ + OSLibDir`, and the latter `/usr/lib` could be removed  if OSLibDir is "lib". 
I am testing https://github.com/llvm/llvm-project/pull/128428 and will push it when the Windows tests pass (it's annoying but Windows uses \\, making a lot of tests ugly..)

---

A lot of ELF-specific Triple::VendorType should not have been added in the first place.

Can you confirm whether you added new OpenEmbedded VendorType due to the following customization?
This is an unfortunate special case and I hope that OpenEmbedded moves away from the difference. Just pick gcc/ or gcc-cross/ or install symlinks so that Clang doesn't need to more probing.
It does not scale if there are similar projects like OpenEmbedded that do a similar thing and want to add a new VendorType to LLVM just to make their upstream divergent customization working.

```
      // This is the normal place.
      {"gcc/" + CandidateTriple.str(), "../..", GCCDirExists},

      // Debian puts cross-compilers in gcc-cross.
      {"gcc-cross/" + CandidateTriple.str(), "../..", GCCCrossDirExists},

      // The Freescale PPC SDK has the gcc libraries in
      // <sysroot>/usr/lib/<triple>/x.y.z so have a look there as well. Only do
      // this on Freescale triples, though, since some systems put a *lot* of
      // files in that location, not just GCC installation data.
      {CandidateTriple.str(), "..",
       TargetTriple.getVendor() == llvm::Triple::Freescale ||
           TargetTriple.getVendor() == llvm::Triple::OpenEmbedded}};
```

mikecrowe commented:
Thank you for investigating this.  Your change in https://github.com/llvm/llvm-project/pull/128428 certainly makes things look a lot neater to me. I believe that OpenEmbedded multilib is relying on the same `llvm::Triple::OpenEmbedded` workaround added in https://reviews.llvm.org/D48862.

I rebased this change on top of current main and the tests I added still need the new cases I added to `Triple.cpp` as would be expected.

I can certainly try asking on the OpenEmbedded lists whether they'd be willing to try to use the more-standard locations. I suspect that they made their decision a very long time ago when the conventions hadn't been settled. The way embedded development is, even if they do agree to change it there will be many users still using older versions of OpenEmbedded without the change for many years to come.

mikecrowe commented:
I [tried asking](https://lists.openembedded.org/g/openembedded-core/topic/gcc_library_directory_and/111849200) about this on the OpenEmbedded mailing list. The response was that it has been this way for so long that it's going to be a major undertaking to change it even though they would quite like to do so.

I had a go at using symlinks to work around the problem. Unfortunately because `ScanLibDirForGCCTriple` ends up using `../../..` to turn `/usr/lib/gcc/${TARGET_SYS}/${gccversion}` (where `${TARGET_SYS}` is the triple) back into just `/usr/lib` it's necessary to have real directories for each of those components and only have symlinks below that. In other words:
```sh
mkdir -p $sysroot/usr/lib/gcc/arm-oemllib32-linux-gnueabi/9.5.0
cd $sysroot/usr/lib/gcc/arm-oemllib32-linux-gnueabi/9.5.0
ln -s ln -s ../../../arm-oemllib32-linux-gnueabi/9.5.0/* .
```

Presumably there's a reason that `../../..` (made up of the `ReversePath` in the `Suffixes` array and another `/../` when constructing `GCCParentLibPath`) is used rather than removing components from the path or just using the `--sysroot` path to find `/usr/lib`?

MaskRay commented:
> I [tried asking](https://lists.openembedded.org/g/openembedded-core/topic/gcc_library_directory_and/111849200) about this on the OpenEmbedded mailing list. The response was that it has been this way for so long that it's going to be a major undertaking to change it even though they would quite like to do so.
> 
> I had a go at using symlinks to work around the problem. Unfortunately because `ScanLibDirForGCCTriple` ends up using `../../..` to turn `/usr/lib/gcc/${TARGET_SYS}/${gccversion}` (where `${TARGET_SYS}` is the triple) back into just `/usr/lib` it's necessary to have real directories for each of those components and only have symlinks below that. In other words:
> 
> ```shell
> mkdir -p $sysroot/usr/lib/gcc/arm-oemllib32-linux-gnueabi/9.5.0
> cd $sysroot/usr/lib/gcc/arm-oemllib32-linux-gnueabi/9.5.0
> ln -s ln -s ../../../arm-oemllib32-linux-gnueabi/9.5.0/* .
> ```
> 
> Presumably there's a reason that `../../..` (made up of the `ReversePath` in the `Suffixes` array and another `/../` when constructing `GCCParentLibPath`) is used rather than removing components from the path or just using the `--sysroot` path to find `/usr/lib`?

Thank you for asking OpenEmbedded.
clangDriver works with many toolchains. The issue is that if each one require its own special handling in the code, it will add up, and it's definitely not a scalable solution...
Ideally, the GCC code would introduce symlinks, eliminating the need for clangDriver to handle these variations. This would prevent situations where clangDriver cleanup requires OpenEmbedded users to manually verify potential breakages.
